pvlib/bifacial/infinite_sheds.py

```diff
@@ -495,6 +501,16 @@ def get_irradiance_poa(surface_tilt, surface_azimuth, solar_zenith,
     --------
     get_irradiance
     """
+    if model == 'haydavies':
+        if dni_extra is None:
+            raise ValueError(f'must supply dni_extra for {model} model')
+        # call haydavies function and request components to help adjust dni/dhi
+        sky_diffuse_components = haydavies(0, 180, dhi, dni, dni_extra,
+                                           solar_zenith, solar_azimuth,
+                                           return_components=True)
+        dhi = dhi - sky_diffuse_components['circumsolar']
+        dni = (ghi - dhi) / cosd(solar_zenith)
```

Comment:
> DHI to >=0

This should be unnecessary, since `sky_diffuse_components['circumsolar'] < dhi` in the Hay-Davies model.

I think something like `min_cos_zenith` is needed. An alternative: what about a second call to `haydavies` to get the circumsolar on the normal plane?